### DISCUSSION TITLE
Does Pioneer support GraphQL file upload?

### DISCUSSION THREAD

**Discussion by d-exclaimation** [d-exclaimation/pioneer#103] [Q&A] [answered]
Does Pioneer support [GraphQL File Upload](https://github.com/jaydenseric/graphql-multipart-request-spec#server)? Why or why not?

**Comment by d-exclaimation** [comment, marked as answer]
Pioneer does not have built-in support for [GraphQL File Upload](https://github.com/jaydenseric/graphql-multipart-request-spec#server). 

Pioneer took many similar approaches and even some intentional design decisions to `@apollo/server`. GraphQL file upload isn't part of `@apollo/server` since `v3` due to some security concerns allowing `multipart/form-data` particularly [CSRF Attacks](https://pioneer-graphql.netlify.app/features/graphql-over-http/#csrf-and-xs-search).

In addition, this [blog post](https://www.apollographql.com/blog/backend/file-uploads/file-upload-best-practices/) somewhat explained the flaws in having file upload to be as part of GraphQL server and other better alternatives for doing file uploads with GraphQL.

In short, it is suggested that we [_"Separate the GraphQL Layer from the Service Layer"_](https://principledgraphql.com/operations#10-separate-the-graphql-layer-from-the-service-layer).
